types/src/dkg/real_dkg/rounding/mod.rs

```diff
@@ -199,51 +229,61 @@ impl DKGRoundingProfile {
             }
         }
 
-        // todo: remove once aptos-dkg supports 0 weights
-        if !is_valid_profile(&best_profile, reconstruct_threshold_in_stake_ratio) {
-            println!("[Randomness] Rounding error: failed to find a valid profile, using default");
-            return Self::default(
-                validator_stakes.len(),
-                secrecy_threshold_in_stake_ratio,
-                reconstruct_threshold_in_stake_ratio,
+        if is_valid_profile(&best_profile, reconstruct_threshold_in_stake_ratio) {
+            Ok(best_profile)
+        } else {
+            bail!(
+                "could not find a valid weight in the given weight range [{}, {}]",
+                total_weight_min,
+                total_weight_max
             );
         }
-
-        best_profile
     }
 
-    pub fn default(
-        num_validators: usize,
-        secrecy_threshold_in_stake_ratio: U64F64,
-        reconstruct_threshold_in_stake_ratio: U64F64,
+    pub fn infallible(
+        validator_stakes: &Vec<u64>,
+        mut secrecy_threshold_in_stake_ratio: U64F64,
+        mut reconstruct_threshold_in_stake_ratio: U64F64,
+        fast_secrecy_threshold_in_stake_ratio: Option<U64F64>,
     ) -> Self {
-        Self {
-            validator_weights: vec![1; num_validators],
+        let one = U64F64::from_num(1);
+        secrecy_threshold_in_stake_ratio = min(one, secrecy_threshold_in_stake_ratio);
+        reconstruct_threshold_in_stake_ratio = min(one, reconstruct_threshold_in_stake_ratio);
+        reconstruct_threshold_in_stake_ratio = max(
             secrecy_threshold_in_stake_ratio,
             reconstruct_threshold_in_stake_ratio,
-            reconstruct_threshold_in_weights: (U64F64::from_num(num_validators)
-                * secrecy_threshold_in_stake_ratio)
-                .to_num::<u64>(),
-            fast_reconstruct_threshold_in_stake_ratio: None,
-            fast_reconstruct_threshold_in_weights: None,
-        }
+        );
+
+        let stake_total = U64F64::from_num(validator_stakes.clone().into_iter().sum::<u64>());
+
+        // This `stake_per_weight` value guarantees a weight assignment that achieves liveness and privacy
+        let stake_per_weight = stake_total
+            * 2
+            * (reconstruct_threshold_in_stake_ratio - secrecy_threshold_in_stake_ratio)
+            / U64F64::from_num(validator_stakes.len())
+            - U64F64::from_num(1);
```

Comment:
Why -1 here? can it be negative?

Reply:
right, unlikely in practice but bad input can lead to negative...

Reply:
fixed to use the infallable `total_weight_upper_bound`